src/poetry/console/commands/check.py

```diff
@@ -88,6 +91,44 @@ def _validate_readme(self, readme: str | list[str], poetry_file: Path) -> list[s
                 errors.append(f"Declared README file does not exist: {name}")
         return errors
 
+    def _validate_dependencies_source(self, config: dict[str, Any]) -> list[str]:
+        """Check dependencies's source are valid"""
+        dependency_sources: dict[str, set[str]] = defaultdict(set)
+        sources = {k["name"] for k in config.get("source", [])}
+        errors = []
+
+        dependency_declarations: list[
+            dict[str, str | dict[str, str] | list[dict[str, str]]]
+        ] = []
+        # scan dependencies and group dependencies settings in pyproject.toml
+        if "dependencies" in config:
+            dependency_declarations.append(config["dependencies"])
+
+        for group in config.get("group", {}).values():
+            if "dependencies" in group:
+                dependency_declarations.append(group["dependencies"])
+
+        for dependency_declaration in dependency_declarations:
+            for dependency, declaration in dependency_declaration.items():
+                if isinstance(declaration, list):
+                    for item in declaration:
+                        if "source" in item:
+                            dependency_sources[dependency].add(item["source"])
+                elif isinstance(declaration, dict) and "source" in declaration:
+                    dependency_sources[dependency].add(declaration["source"])
+
+        all_referenced_sources: set[str] = reduce(
+            lambda i, j: i | j, dependency_sources.values(), set()
+        )
```

Comment:
Why not build `all_referenced_sources` directly and make a detour by `dependency_sources`?

Reply:
good point, I should build all_referenced_sources directly